api/C/include/csi_defs.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * General-Purpose Definitions for use in CSI code
+ *
+ * This file is included from rvm_csi.h prior to all other headers and contains
+ * general-purpose definitions which may be used by subsequent headers.
+ *
+ * Copyright (c) RISC-V International 2022. Creative Commons License. Auto-
+ * generated file: DO NOT EDIT
+ */
+
+#ifndef CSI_DEFS_H
+#define CSI_DEFS_H
+
+/*
+ * Return type for functions indicating status
+ */
+typedef enum {
+    CSI_SUCCESS = 0,
+    CSI_NOT_IMPLEMENTED = -1,
+    CSI_ERROR = -2,
+    CSI_NOT_INITIALIZED = -3,
+    CSI_OUT_OF_MEM = -4,
+} csi_status_t;
+
+/*
+ * Enumerates all sources in the system that can give rise to a trap event.  This
+ * enumeration should be treated as an integer which can have additional values
+ * beyond those enumerated here.  A standard set of interrupt and exception sources
+ * are enumerated here, which exclude external sources.  All external interrupt
+ * sources should be enumerated by the BSP within csi_bsp_interrupts.h; and their
+ * ennumerated values should follow on from CSI_NUM_STANDARD_TRAP_SOURCES.
+ * csi_bsp_interrupts.h must also define CSI_TOTAL_BSP_TRAP_SOURCES to indicate the
+ * total number of sources defined.  User applications may then use values above
+ * this to enumerate user-defined software signals.  Note that the actual
+ * enumerated values have no meaning beyond serving to uniquely identify the source
+ * of a trap event.
+ */
+typedef enum {
+    CSI_ENUM_NMI = 0,
+    CSI_ENUM_MACHINE_SW_INTERRUPT,
+    CSI_ENUM_MACHINE_TIMER_INTERRUPT,
+    CSI_ENUM_INSTRUCTION_ADDR_MISALIGNED_EXCEPTION,
+    CSI_ENUM_INSTRUCTION_ACCESS_FAULT_EXCEPTION,
+    CSI_ENUM_ILLEGAL_INSTRUCTION_EXCEPTION,
+    CSI_ENUM_BREAKPOINT_EXCEPTION,
+    CSI_ENUM_LOAD_ADDR_MISALIGNED_EXCEPTION,
+    CSI_ENUM_LOAD_ACCESS_FAULT_EXCEPTION,
+    CSI_ENUM_STORE_ADDR_MISALIGNED_EXCEPTION,
+    CSI_ENUM_STORE_ACCESS_FAULT_EXCEPTION,
+    CSI_ENUM_ECALL_FROM_UMODE,
+    CSI_ENUM_ECALL_FROM_MMODE,
+    CSI_ENUM_INST_PAGE_FAULT,
+    CSI_ENUM_LOAD_PAGE_FAULT,
+    CSI_ENUM_STORE_PAGE_FAULT,
+    CSI_NUM_STANDARD_TRAP_SOURCES, /* Must come last in this list */
+} csi_trap_source_t;
```

Comment:
Should trap source classified to interrupt and exception to match mcause.

![image](https://user-images.githubusercontent.com/1538922/218364737-bb333f1a-0008-446e-8dfc-87cc94c619fd.png)

This is how we define in NMSIS Core https://github.com/Nuclei-Software/NMSIS/blob/master/Device/_Template_Vendor/Vendor/Device/Include/Device.h#L64-L114

Reply:
csi_trap_source_t is different from mcause, because it is much more detailed.  For example, mcause only has a single entry for "machine external interrrupt", whereas csi_trap_source_t has to enumerate all the external interrupts in the system (the BSP extends it for this purpose).  Also, it is useful to have all values of csi_trap_source_t contiguous (without gaps), so that the BSP can maintain a table  which is indexed by this enumeration (without wasting memory).

I think there are no efficiency problems here because a decode from mcause (+ PLIC/AIA configuration) to csi_trap_source_t does not need to happen at the time an interrupt is taken (it can happen at the time a handler is registered).  csi_trap_source_t just gives a convenient way of indexing sources for the purpose of the HAL API.

Reply:
> Also, it is useful to have all values of csi_trap_source_t contiguous (without gaps), so that the BSP can maintain a table which is indexed by this enumeration (without wasting memory)

Regarding the contiguous values, I think exception and interrupt should be enumerated in two enum, since exception id and interrupt id could be the same.

Reply:
Sorry I think we have a misunderstanding.  The way I conceived it, the actual enumerated values are not important, they don't relate to anything in the hardware.  We have a single function (csi_register_m_isr) to register a (machine-mode) user handler for either interrupts or exceptions, so need a single enum.

Reply:
OK, so all the handing of the `csi_trap_source_t` will leave to the implementation of  csi_register_m_isr to process this enum.

Reply:
And did you take a look at the riscv fast interrupt [1] for mcu, it support vector interrupt and non-vector interrupt, it might be more necessary than plic/aia.

[1]: https://github.com/riscv/riscv-fast-interrupt/blob/master/clic.adoc